src/main/java/uk/gov/ons/ctp/response/action/service/ProcessEventService.java

```diff
@@ -0,0 +1,766 @@
+  }
+
+  /**
+   * Processes Events. This method takes two attributes collection Exercise Id and Event Tag.
+   * GetsCollectionExerciseDTO and SurveyDTO from @param collectionExerciseId Gets all email cases
+   * against collectionExerciseID (isActiveEnrolment == true) Maps email cases to ActionTemplate
+   * processes email cases if actionable Gets all letter cases against collectionExerciseID
+   * (isActiveEnrolment == false) Maps letter cases to ActionTemplate processes letter cases if
+   * actionable
+   *
+   * @param collectionExerciseId - is passed by collectionexercisesvc scheduler as a part of the
+   *     event process.
+   * @param eventTag - is passed by collectionexercisesvc scheduler as a part of the event process.
+   */
+  @Async
+  public void processEvents(UUID collectionExerciseId, String eventTag) {
+    log.with("collectionExerciseId", collectionExerciseId)
+        .with("event_tag", eventTag)
+        .info("Started processing");
+    Instant instant = Instant.now();
+    CollectionExerciseDTO collectionExercise = getCollectionExercise(collectionExerciseId);
+    SurveyDTO survey = getSurvey(collectionExercise.getSurveyId());
+    log.debug("Getting Email cases against collectionExerciseId and event active enrolment");
+    List<ActionCase> email_cases =
+        actionCaseRepository.findByCollectionExerciseIdAndActiveEnrolment(
+            collectionExerciseId, true);
+    if (email_cases.size() > 0) {
+      log.with("collectionExerciseId", collectionExercise).info("Processing email cases.");
+      ActionTemplate actionTemplate = actionTemplateService.mapEventTagToTemplate(eventTag, true);
+      if (actionTemplate != null) {
+        log.with("email cases", email_cases.size()).info("Processing email cases");
+        email_cases
+            .parallelStream()
+            .filter(c -> isActionable(c, actionTemplate))
+            .forEach(
+                c ->
+                    processEmailCase(c, collectionExercise, survey, actionTemplate, null, instant));
+      } else {
+        log.with("activeEnrolment", true)
+            .with("event", eventTag)
+            .error("No Template registered against the event and active enrolment");
+      }
+    } else {
+      log.with("collectionExercise", collectionExercise)
+          .with("event", eventTag)
+          .info("No Emails to process");
+    }
+    log.debug("Getting Email cases against collectionExerciseId and event active enrolment");
+    List<ActionCase> letter_cases =
+        actionCaseRepository.findByCollectionExerciseIdAndActiveEnrolment(
+            collectionExerciseId, false);
+    if (letter_cases.size() > 0) {
+      ActionTemplate actionTemplate = actionTemplateService.mapEventTagToTemplate(eventTag, false);
+      if (actionTemplate != null) {
+        log.with("Letter cases", letter_cases.size()).info("Processing letter cases");
+        processLetterCases(letter_cases, actionTemplate, survey, collectionExercise, instant, true);
```

Comment:
It's a little weird that we filter email_cases in this method but filter letter_cases in the method we're calling. It'd be good to have some consistency; either stream both types of case here, or stream both types of case in processLetterCases/processEventCases.

Reply:
yes sure it can be consistent. Will extract it as a method.